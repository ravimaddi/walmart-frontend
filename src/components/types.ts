export interface User {
  id: number
  name: string
  email: string
  phone: string
  website: string
  address: Address
}

interface Address {
  city: string
}

export interface post {
  userId: number
  id: number
  title: string
  body: string
}

export interface album {
  userId: number
  title: string
  id: number
}

export interface photo {
  albumId: number
  id: number
  title: string
  url: string
  thumbnailUrl: string
}
