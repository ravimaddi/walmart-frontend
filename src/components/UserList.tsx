import React, { useState } from "react"
import { useNavigate } from "react-router-dom"
import { useAxios } from "../hooks"
import { User } from "./types"
import {
  Container,
  Stack,
  Popover,
  Typography,
  List,
  ListItemText,
  ListItemButton,
  CircularProgress,
  Alert,
  AlertTitle,
} from "@mui/material"

export const UserList: React.FC = () => {
  const {
    data: users,
    error,
    loading,
  } = useAxios({
    url: "users",
  })
  const [popoverContent, setPopoverContent] = useState(null)
  const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null)
  const navigate = useNavigate()

  if (loading) {
    return <CircularProgress />
  }
  if (error) {
    return (
      <Alert severity="error">
        <AlertTitle>Error</AlertTitle>
        This is an error alert â€” <strong>check it out!</strong>
      </Alert>
    )
  }

  const handlePopoverOpen = (
    event: React.MouseEvent<HTMLElement>,
    user: User
  ) => {
    setAnchorEl(event.currentTarget)
    setPopoverContent(user)
  }

  const handlePopoverClose = () => {
    setAnchorEl(null)
  }
  const open = Boolean(anchorEl)
  return (
    <Container>
      <h1>Users List</h1>
      <Stack spacing={2}>
        {users?.map((user: User) => (
          <List
            key={user?.id}
            onMouseEnter={(e) => handlePopoverOpen(e, user)}
            onMouseLeave={handlePopoverClose}
          >
            <ListItemButton onClick={() => navigate(`user/${user?.id}`)}>
              <ListItemText primary={user?.name} />
            </ListItemButton>
          </List>
        ))}
      </Stack>
      <Popover
        id="mouse-over-popover"
        sx={{
          pointerEvents: "none",
        }}
        open={open}
        anchorEl={anchorEl}
        anchorOrigin={{
          vertical: "bottom",
          horizontal: "left",
        }}
        transformOrigin={{
          vertical: "top",
          horizontal: "left",
        }}
        onClose={handlePopoverClose}
        disableRestoreFocus
      >
        <Typography sx={{ p: 1 }}>
          <span>{popoverContent?.username}</span>
          <br />
          <span>{popoverContent?.email}</span>
        </Typography>
      </Popover>
    </Container>
  )
}
