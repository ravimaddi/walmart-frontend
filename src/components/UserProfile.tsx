import { FC } from "react"
import {
  Card,
  CardContent,
  Avatar,
  Container,
  CardHeader,
  List,
  ListItem,
  ListItemText,
  Divider,
  ListItemIcon,
} from "@mui/material"
import { User } from "./types"
import CallIcon from "@mui/icons-material/Call"
import LocationCityIcon from "@mui/icons-material/LocationCity"
import LanguageIcon from "@mui/icons-material/Language"

interface UserProfileProps {
  user: User
}

const style = {
  width: "100%",
  maxWidth: 360,
  bgcolor: "background.paper",
}

export const UserProfile: FC<UserProfileProps> = ({ user }) => {
  return (
    <>
      <Container maxWidth="sm">
        <Card sx={{ maxWidth: "100%" }}>
          <CardHeader
            avatar={
              <Avatar sx={{ bgcolor: "red" }} aria-label="recipe">
                {user?.name[0]}
              </Avatar>
            }
            title={user?.name}
            subheader={user?.email}
          />
          <CardContent>
            <List sx={style}>
              <ListItem button>
                <ListItemIcon>
                  <CallIcon />
                </ListItemIcon>
                <ListItemText primary={user?.phone} />
              </ListItem>
              <Divider />
              <ListItem button>
                <ListItemIcon>
                  <LocationCityIcon />
                </ListItemIcon>
                <ListItemText primary={user?.address?.city} />
              </ListItem>
              <Divider />
              <ListItem button>
                <ListItemIcon>
                  <LanguageIcon />
                </ListItemIcon>
                <ListItemText primary={user?.website} />
              </ListItem>
              <Divider />
            </List>
          </CardContent>
        </Card>
      </Container>
    </>
  )
}
