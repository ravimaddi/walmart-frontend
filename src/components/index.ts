export { UserList } from "./UserList"
export { UserDetails } from "./UserDetails"
export { UserProfile } from "./UserProfile"
