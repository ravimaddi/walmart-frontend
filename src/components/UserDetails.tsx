import { FC } from "react"
import { useParams } from "react-router-dom"
import { useAxios } from "../hooks"
import { UserProfile } from "./UserProfile"
import { UserPosts } from "./posts"
import { AlbumList } from "./albums"
import "../styles/styles.scss"

interface UserDetailsProps {
  [key: string]: any
}

export const UserDetails: FC<UserDetailsProps> = () => {
  const { id } = useParams()
  const { data: user } = useAxios({
    url: `users/${id}`,
  })

  return (
    <div className="container">
      <div className="userDetailsContainer">
        <UserProfile user={user} />
      </div>
      <div className="userPostsContainer">
        <UserPosts userId={id} />
      </div>
      <div className="userAlbumsContainer">
        <AlbumList userId={id} />
      </div>
    </div>
  )
}
