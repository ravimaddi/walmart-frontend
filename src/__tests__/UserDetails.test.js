import { UserDetails } from "../components"
import ShallowRenderer from "react-test-renderer/shallow"

it("check for the userPorfile component", () => {
  const renderer = new ShallowRenderer()
  const view = renderer.render(<UserDetails />)
  expect(view).toMatchSnapshot()
})
