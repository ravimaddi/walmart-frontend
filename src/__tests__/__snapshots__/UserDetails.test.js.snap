// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check for the userPorfile component 1`] = `
<div
  className="container"
>
  <div
    className="userDetailsContainer"
  >
    <UserProfile />
  </div>
  <div
    className="userPostsContainer"
  >
    <UserPosts />
  </div>
  <div
    className="userAlbumsContainer"
  >
    <AlbumList />
  </div>
</div>
`;
